### PR TITLE
Use `DefaultAWSCredentialsProviderChain` AWS SDK class for credentials

### LINKED ISSUE #19556: Use `DefaultAWSCredentialsProviderChain` AWS SDK class for credentials
Follow up discussion at https://github.com/elastic/elasticsearch/pull/18690#issuecomment-234505083

Reading the best practices [recommended by AWS](http://docs.aws.amazon.com/java-sdk/latest/developer-guide/credentials.html), we should use `DefaultAWSCredentialsProviderChain` instead of providing the detail of the chain ourselves.

For now, we read credentials (if not provided in `elasticsearch.yml`) using:

```java
credentials = new AWSCredentialsProviderChain(
   new SystemPropertiesCredentialsProvider(),
   new EnvironmentVariableCredentialsProvider(),
   new InstanceProfileCredentialsProvider()
);
```

Which means that we read from:

* Environment Variables - `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` (RECOMMENDED since they are recognized by all the AWS SDKs and CLI except for .NET), or `AWS_ACCESS_KEY` and `AWS_SECRET_KEY` (only recognized by Java SDK)
* Java System Properties - `aws.accessKeyId` and `aws.secretKey`
* Instance profile credentials delivered through the Amazon EC2 metadata service

Using instead:

```java
credentials = new DefaultAWSCredentialsProviderChain();
```

Will give us two new more methods out of the box:

> * Credential profiles file at the default location (`~/.aws/credentials`) shared by all AWS SDKs and the AWS CLI
> * Credentials delivered through the Amazon EC2 container service if `AWS_CONTAINER_CREDENTIALS_RELATIVE_URI` environment variable is set and security manager has permission to access the variable

Cherry on the cake: as soon as AWS SDK will propose a new implementation, we will benefit from it without any modification (just updating the SDK).

### PR DESCRIPTION
**Pull request by dadoonet**
Follow up discussion at https://github.com/elastic/elasticsearch/pull/18690#issuecomment-234505083

Reading the best practices [recommended by AWS](http://docs.aws.amazon.com/java-sdk/latest/developer-guide/credentials.html), we should use `DefaultAWSCredentialsProviderChain` instead of providing the detail of the chain ourselves.

For now, we read credentials (if not provided in `elasticsearch.yml`) using:

``` java
credentials = new AWSCredentialsProviderChain(
   new SystemPropertiesCredentialsProvider(),
   new EnvironmentVariableCredentialsProvider(),
   new InstanceProfileCredentialsProvider()
);
```

Which means that we read from:
- Environment Variables - `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` (RECOMMENDED since they are recognized by all the AWS SDKs and CLI except for .NET), or `AWS_ACCESS_KEY` and `AWS_SECRET_KEY` (only recognized by Java SDK)
- Java System Properties - `aws.accessKeyId` and `aws.secretKey`
- Instance profile credentials delivered through the Amazon EC2 metadata service

Using instead:

``` java
credentials = new DefaultAWSCredentialsProviderChain();
```

Will give us two new more methods out of the box:

> - Credential profiles file at the default location (`~/.aws/credentials`) shared by all AWS SDKs and the AWS CLI
> - Credentials delivered through the Amazon EC2 container service if `AWS_CONTAINER_CREDENTIALS_RELATIVE_URI` environment variable is set and security manager has permission to access the variable

Cherry on the cake: as soon as AWS SDK will propose a new implementation, we will benefit from it without any modification (just updating the SDK).

We also simplify

```
new AWSCredentialsProviderChain(new StaticCredentialsProvider(new BasicAWSCredentials(key, secret)));
```

As there is no need to wrap StaticCredentialsProvider in AWSCredentialsProviderChain.

Closes #19556.

### CODE REVIEW

**Comment by dadoonet** [comment]
@rjernst I think I find a better solution with this PR to test that we use the right part of the `if/else` code:

```java
        if (key.isEmpty() && secret.isEmpty()) {
            credentials = new DefaultAWSCredentialsProviderChain();
        } else {
            credentials = new StaticCredentialsProvider(new BasicAWSCredentials(key, secret));
        }
```

I'm basically testing here that we are getting same credentials as `DefaultAWSCredentialsProviderChain` would return.
I'm not testing the internal behavior of `DefaultAWSCredentialsProviderChain` as it's not the goal here.

WDYT?

**Comment by dadoonet** [comment]
Well. Actually I found a much more easier way to achieve the goal without needing providing sys props...

File: plugins/discovery-ec2/src/test/java/org/elasticsearch/cloud/aws/AwsEc2ServiceImplTests.java [hunk 1]
```diff
@@ -23,22 +23,22 @@
 import com.amazonaws.Protocol;
 import com.amazonaws.auth.AWSCredentials;
 import com.amazonaws.auth.AWSCredentialsProvider;
+import com.amazonaws.auth.DefaultAWSCredentialsProviderChain;
+import com.amazonaws.util.StringUtils;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.test.ESTestCase;
 
+import static com.amazonaws.SDKGlobalConfiguration.ACCESS_KEY_SYSTEM_PROPERTY;
```
Comment (ywelsch) [inline plugins/discovery-ec2/src/test/java/org/elasticsearch/cloud/aws/AwsEc2ServiceImplTests.java]: import not needed.

File: plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/AwsS3ServiceImplTests.java [hunk 2]
```diff
@@ -23,22 +23,31 @@
...
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.repositories.s3.S3Repository;
 import org.elasticsearch.test.ESTestCase;
 
+import static com.amazonaws.SDKGlobalConfiguration.ACCESS_KEY_ENV_VAR;
+import static com.amazonaws.SDKGlobalConfiguration.ACCESS_KEY_SYSTEM_PROPERTY;
+import static com.amazonaws.SDKGlobalConfiguration.ALTERNATE_ACCESS_KEY_ENV_VAR;
+import static com.amazonaws.SDKGlobalConfiguration.ALTERNATE_SECRET_KEY_ENV_VAR;
+import static com.amazonaws.SDKGlobalConfiguration.SECRET_KEY_ENV_VAR;
+import static com.amazonaws.SDKGlobalConfiguration.SECRET_KEY_SYSTEM_PROPERTY;
```
Comment (ywelsch) [inline plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/AwsS3ServiceImplTests.java]: imports not needed.

**Comment by ywelsch** [comment]
LGTM on the change. We should (not directly related to the change in this PR) also add tests that check the various credential providers and show what the user needs to do to make them work with the security manager.

**Comment by dadoonet** [comment]
Thank you @ywelsch!

**Pull request merged**